### ISSUE TITLE
[HOLD for payment 2023-07-21] [$2000] Selecting a chat message and pressing any key focus the message instead of switching the focus to the composer.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#16098]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1.  Go to any chats
2.  Click on any message
3.  Press any letter.

## Expected Result:
Character is typed in the message composer

## Actual Result:
Message gets focused instead

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.2.87-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/226144582-9575020a-f3c2-48e3-a7a4-82c2f47cfcdb.mp4

https://user-images.githubusercontent.com/43996225/226144598-adbaa189-7e4c-41da-94a3-50736a43c410.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @pecanoro 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1679075457783789

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0108a1ba7c41369017</li>
        <li>Upwork Job ID: 1640319237704986624</li>
        <li>Last Price Increase: 2023-06-19</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sophiepintoraetz (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by sophiepintoraetz** [comment]
Tomorrow

**Comment by sophiepintoraetz** [comment]
Hopefully have some time today - now that I'm recovering!

**Comment by MelvinBot** [comment]
Triggered auto assignment to @PauloGasparSv (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by PauloGasparSv** [comment]
Making this external!

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0108a1ba7c41369017

**Comment by MelvinBot** [comment]
Current assignee @sophiepintoraetz is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`External`)

**Comment by MelvinBot** [comment]
Current assignee @PauloGasparSv is eligible for the External assigner, not assigning anyone new.

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When clicking on a chat message and click on any character, the message gets focused and the composer is not.

### What is the root cause of that problem?
The `ReportActionItem` is focusable, and we don't have any `keydown` listener to focus the composer, so the `ReportActionItem` will get focused.

### What changes do you think we should make in order to solve the problem?

We should make the behavior consistent with Slack, by automatically focusing on the Composer if we press on regular characters like a, b, c, d (not non-character keys like `Enter`, `Shift`) when we're not typing on any input or textarea.

This can be done by:
1. Add a new `keypressListener ` listener method to the `ReportActionCompose` [here](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionCompose.js)
```
keypressListener(e) {
        if (this.state.isFocused) return;

        // if the key pressed is non-character keys like Enter, Shift, ... do not focus
        if (e.key.length > 1) return;

        // if we're typing on another input/text area, do not focus
        if (e.target.nodeName === 'INPUT' || e.target.nodeName === 'TEXTAREA') return;

        // Otherwise, focus the composer and update the comment with the pressed key
        this.focus();
        this.updateComment(`${this.state.value}${e.key}`, true);
    }
```
and bind it in the constructor

2. In `componentDidMount` [here](https://github.com/Expensify/App/blob/f3878b919b1a1d52988ef033914f6c66de8338b6/src/pages/home/report/ReportActionCompose.js#L202) add `document.addEventListener('keydown', this.keypressListener);`
(We shouldn't use `keypress` because it's a deprecated method, also it will still capture keys like `Enter` so there's not much difference from `keydown` in our use case)

3. Add `document.removeEventListener('keydown', this.keypressListener);` to `componentWillUnmount`.

### What alternative solutions did you explore? (Optional)
- If we want to be more precise with the `non-character key` check (` if (e.key.length > 1) return;`) we can go with a regex as well.
- We need to decide what we want to do for Message edit case since there could be multiple edit messages opening at once. Slack only auto focuses on the main composer, not the edit composer. For that the above solution is enough.

**Comment by PauloGasparSv** [comment]
Not overdue!

**Comment by rushatgabhane** [comment]
@getusha @tienifr your solutions are code diffs. Please explain your solution in plain english

**Comment by rushatgabhane** [comment]
@PauloGasparSv @sophiepintoraetz I'm curious about what exactly we want here

Precondition - composer is not focused.
1. Pressing any key will focus the composer
2. Or, pressing any key will focus the composer and insert it in the composer as well

**Comment by rushatgabhane** [comment]
Asked @pecanoro on [slack](https://expensify.slack.com/archives/C049HHMV9SM/p1680230004526819?thread_ts=1679075457.783789&cid=C049HHMV9SM) to post a problem solution because I believe this is not a bug, but a feature request.

**Comment by PauloGasparSv** [comment]
> Asked @pecanoro on [slack](https://expensify.slack.com/archives/C049HHMV9SM/p1680230004526819?thread_ts=1679075457.783789&cid=C049HHMV9SM) to post a problem solution because I believe this is not a bug, but a feature request.

Thanks for that @rushatgabhane let's wait for that response

> Or, pressing any key will focus the composer and insert it in the composer as well

That's a great question, I don't think that's defined here. I see you asked that on slack too : )
IMO that is the intent and we should only do it for alphanumeric keys.

**Comment by sophiepintoraetz** [comment]
Commented on the slack thread but let's not draw this process out unnecessarily with a P/S as this is a bug not a feature request.

**Comment by MelvinBot** [comment]
@PauloGasparSv @rushatgabhane @sophiepintoraetz this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by rushatgabhane** [comment]
> but let's not draw this process out unnecessarily with a P/S

Okayy sounds good.

To move this forward, let's just go with (2) unless someone disagrees
- pressing any key will focus the composer and insert it in the composer as well.

**Comment by tienifr** [comment]
> @getusha @tienifr the solutions are code diffs. We don't accept that now. Please update the proposals by explaining the solution in plain english. Thanks 🙇

@rushatgabhane I've updated [the proposal](https://github.com/Expensify/App/issues/16098#issuecomment-1485325162) to avoid code diffs, thanks!

**Comment by PauloGasparSv** [comment]
Hey @rushatgabhane  I like @tienifr's updated proposal! Can you take a look at it (curious on your feedback)?

**Comment by rushatgabhane** [comment]
on it

**Comment by rushatgabhane** [comment]
@PauloGasparSv I recommend @tienifr's [proposal](https://github.com/Expensify/App/issues/16098#issuecomment-1485325162)

Pressing any key will focus the composer and insert it in the composer wherever the previous selection was placed.

**Comment by rushatgabhane** [comment]
@getusha's proposal appends the value to composer instead of placing it wherever the previous selection (cursor) was

**Comment by MelvinBot** [comment]
📣 @tienifr You have been assigned to this job by @PauloGasparSv!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by rushatgabhane** [comment]
@PauloGasparSv we have to undo price to 1k 
 sorry for delay!

**Comment by PauloGasparSv** [comment]
Nice catch @rushatgabhane! The approved proposal predates the change so let me rollback to $1000

cc @sophiepintoraetz

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $1000

**Comment by tienifr** [comment]
@rushatgabhane  The PR is ready for review: https://github.com/Expensify/App/pull/17138

**Comment by sophiepintoraetz** [comment]
Hey @rushatgabhane, will you be able to review the PR? 🙏

**Comment by tienifr** [comment]
@rushatgabhane Many thank for your fix

**Comment by tienifr** [comment]
# Regression Test Proposal

**Bug**: Selecting a chat message and pressing any key focus the message instead of switching the focus to the composer

**Proposed Test Steps**:

1. Go to any chats
2. Click on any message
3. Press any character keys like a,b,c,d (not non-character keys like Enter, Shift)
4. Verify that the composer is focused again and the above character is entered in the composer
5. Open Search Page and blur the input on the search page
6. Press any character keys like a,b,c,d (not non-character keys like Enter, Shift)
7. Verify that these characters aren't entered in the main composer
8. Open Emoji Picker and blur the input on the search
9.  Press any character keys like a,b,c,d (not non-character keys like Enter, Shift)
10. Verify that these characters aren't entered in the main composer

Do we :+1: or :-1:

**Comment by PauloGasparSv** [comment]
I think we caused another regression reported by @situchan  [here](https://expensify.slack.com/archives/C049HHMV9SM/p1681367247044099)! In the first P.R. we caught a bug and fixed it where the composer box and the emoji search input would receive keyboard input at the same time.

This bug is similar, clicking in the "Change Default Skin Tone" is causing that to happen again for some reason.
I'm investigating here.

**Comment by situchan** [comment]
> I think we caused another regression reported by @situchan [here](https://expensify.slack.com/archives/C049HHMV9SM/p1681367247044099)!

GH created: https://github.com/Expensify/App/issues/17413

**Comment by situchan** [comment]
Another related to issue: https://github.com/Expensify/App/issues/17411

**Comment by PauloGasparSv** [comment]
> Another related to issue: https://github.com/Expensify/App/issues/17411

That one was discussed [here](https://expensify.slack.com/archives/C049HHMV9SM/p1681329759993079) and [here](https://github.com/Expensify/App/issues/17397) and I think it should be treated as a separate issue : )

**Comment by Pujan92** [comment]
Seems regression when in RHN opened popover closes, it focuses the composer even though the RHN is still open.

https://user-images.githubusercontent.com/14358475/232117268-650f1c70-37e5-470b-9227-0ce1250cdf74.mov

**Comment by PauloGasparSv** [comment]
I think we should revert these two, there too many regressions caused by these changes:

- https://github.com/Expensify/App/pull/17360
- https://github.com/Expensify/App/pull/17138

And get rid of https://github.com/Expensify/App/pull/17428 or maybe use it to remove the change if reverting both P.R.'s is a challenge.

**Comment by MelvinBot** [comment]
⚠️ Looks like this issue was linked to a **Deploy Blocker** [here](https://github.com/Expensify/App/issues/17465)

If you are the assigned CME please investigate whether the linked PR caused a regression and leave a comment with the results.

If a regression has occurred and you are the assigned CM follow the instructions [here](https://stackoverflow.com/c/expensify/questions/13647).

If this regression could have been avoided please consider also proposing a recommendation to the PR checklist so that we can avoid it in the future.

**Comment by tienifr** [comment]
@PauloGasparSv @rushatgabhane  Hi sorry for the trouble here, [PR to revert](https://github.com/Expensify/App/pull/17475) is ready.

**Comment by sophiepintoraetz** [comment]
@tienifr or @rushatgabhane - can you summarise the number of regressions [attached to the PR](https://github.com/Expensify/App/pull/17138), please?

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.3.0-2 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/17138
- https://github.com/Expensify/App/pull/17360


If no regressions arise, payment will be issued on 2023-04-24. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@rushatgabhane / @PauloGasparSv] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@rushatgabhane / @PauloGasparSv] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@rushatgabhane / @PauloGasparSv] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@sophiepintoraetz] Determine if we should create a regression test for this bug.
- [ ] [@rushatgabhane] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@sophiepintoraetz] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by sophiepintoraetz** [comment]
So just to be clear here do we have a solution in place to fix the original issue?

**Comment by tienifr** [comment]
we have 2 regression bugs, I am doing on it

**Comment by tienifr** [comment]
### Regression Bugs

There are 2 regression bugs here:

1. [This issue](https://github.com/Expensify/App/issues/17413) 

**RCA**: Because in the previous PR, we only check the emoji picker is open in the composer. There are other places in reportScreen that also trigger to open emojiPicker

**Solution**: We should create a new field in ONYX to keep track of the state of the emoji picker (open or closed) and then use this field to check whether we should listen key-down and focus on the composer. I just created [a new PR](https://github.com/Expensify/App/pull/17645) with all changes of PR's code change reverted and also apply the above solution. So both issues [here](https://github.com/Expensify/App/issues/17413) and [here](https://github.com/Expensify/App/issues/16098) will be resolved

2. [This issue](https://github.com/Expensify/App/issues/17478)

**RCA**: When the second popover modal is closed, it is setting the isVisible of modal in ONYX to false, while the parent popover is still opened -> wrong condition

**Solution 1**:
- Add a new key in Onyx that stores the number of modal being opened: OPENING_MODAL_COUNT: `openingModalCount`
- Inside function `setModalVisibility()` in Modal.js, add check conditions to update ONYXKEYS.OPENING_MODAL_COUNT and ONYXKEYS.MODAL respectively

**Solution 2**:
- Add `shouldSetModalVisibility={false}` in the Modal that has level > 1 (not the first popover)

Cc: @rushatgabhane @sophiepintoraetz @PauloGasparSv

**Comment by tienifr** [comment]
@rushatgabhane @PauloGasparSv @sophiepintoraetz 
> When the second popover modal is closed, it is setting the isVisible of modal in ONYX to false, while the parent popover is still opened -> wrong condition

I think this bug doesn't come from my PR so should we handle it in other issue?

**Comment by tienifr** [comment]
@rushatgabhane @PauloGasparSv Here is my new PR: https://github.com/Expensify/App/pull/17649. It's the combination of https://github.com/Expensify/App/pull/17138 and https://github.com/Expensify/App/pull/17360
Just one minor update: I want to add new condition `EmojiPickerAction.emojiPickerRef.current.state.isEmojiPickerVisible` to prevent key press event when the emoji picker is open. But the isVisible of modal in OnYX doesn't work as expect as I mentioned above, I think we can fix it on other issue. What do you think?

I'm really sorry for this inconvenience. I'll try my best to resolve this issue ASAP. Thanks  🚀 🚀 🚀

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@rushatgabhane] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@rushatgabhane] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@rushatgabhane] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@sophiepintoraetz] Determine if we should create a regression test for this bug.
- [ ] [@rushatgabhane] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@sophiepintoraetz] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by sophiepintoraetz** [comment]
@PauloGasparSv - what's needed here to close out this issue? Are you able to comment on the PR?

**Comment by sophiepintoraetz** [comment]
I've actually also stumbled across this issue here https://github.com/Expensify/App/issues/15992 whilst being assigned to other issues - so I wonder if we fold it into there as well?

**Comment by PauloGasparSv** [comment]
Hey @sophiepintoraetz sorry for the delay here, was focused on N6.9!

Both me and @rushatgabhane feel like we shouldn't implement the feature because of all the edge cases and  regressions it can cause so I started a [slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1682620217672419) to discuss if we should implement this or not (and how to prevent regressions if we decide to implement it).

**Comment by sophiepintoraetz** [comment]
Thanks for starting that conversation, @PauloGasparSv! If I'm reading that right, are we agreeing **not** to do it?

@rushatgabhane looks like I owe you an apology - perhaps a P/S would have prevented some of the downstream issues we faced!

**Comment by PauloGasparSv** [comment]
> If I'm reading that right, are we agreeing not to do it?

Coming from the [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1682620217672419) we agreed that:

1. We shouldn't dismiss this only because of regressions and bugs
2. It's ok if we don't do it for another reason but since WhatsApp and Slack have this feature that's a reason to implement.

IMO since our only concerns were regressions we **should try to implement this again** and **update the Test/QA steps of the P.R so we test all flows that created regressions last time**.
I also think it's a good idea to ask more people to review the [P.R.](https://github.com/Expensify/App/pull/17649) before merging (it maybe by assigning puller bear again).

Does that sound good? cc @sophiepintoraetz @rushatgabhane @tienifr

**Comment by sophiepintoraetz** [comment]
Got it! Yes, I think we seek more reviews on the PR and take a more considered approach when building the PR 🙏

**Comment by sophiepintoraetz** [comment]
@mallenexpensify - I'm about to go OOO for the week so reassigning. There's been a few regressions so I don't think this issue will be eligible for a bonus payment should it come to that while I'm away (possibly a penalty), as Tienfr as assigned on Apr 6 and the PR is yet to merge. We are waiting on @rushatgabhane to review @tienifr's PR - it's been sitting for a week so have [bumped Rushat in Slack](https://expensify.slack.com/archives/C02NK2DQWUX/p1684748601787799).

**Comment by mallenexpensify** [comment]
@rushatgabhane please treat this as a priority, removed `Awaiting Payment` and `Reviewing` so it'll go over due and help light a 🔥

**Comment by PauloGasparSv** [comment]
P.R. still in review!

**Comment by PauloGasparSv** [comment]
P.R. still in review, latest updates [here](https://github.com/Expensify/App/pull/17649#issuecomment-1566545218)

**Comment by PauloGasparSv** [comment]
Not overdue, we are still in progress and discussing a solution for [this bug](https://github.com/Expensify/App/pull/17649#issuecomment-1558947680) in the P.R.

**Comment by sophiepintoraetz** [comment]
Thanks @mallenexpensify for looking after it! I'll take this back now that I'm in office but def let me know if you'd like to stay assigned.

**Comment by sophiepintoraetz** [comment]
How are we tracking here, @PauloGasparSv ?

**Comment by PauloGasparSv** [comment]
>  How are we tracking here, @PauloGasparSv ?

There were no updates this past 2 days, we are still having [this bug](https://github.com/Expensify/App/issues/16098#issuecomment-1570948590).

I could not help with the bug yet but will stop do to so ASAP. Also bumped the contributors in case they have an update [here](https://github.com/Expensify/App/pull/17649#issuecomment-1573879574)

**Comment by sophiepintoraetz** [comment]
@PauloGasparSv @tienifr @rushatgabhane - can you please summarise the bugs/issues that you are stuck on and we'll take it to slack for suggestions.

Alternatively, we can look at assigning an agency to push this along further as well for ideas.

**Comment by PauloGasparSv** [comment]
Well, that didn't work! Asking internally if another C+ can take this over.

**Comment by sophiepintoraetz** [comment]
I am OOO for the next week (back on 20 Jun) so thank you @mallenexpensify for looking after this. This issue has been moving slowly and possibly needs some help with the external agencies. The contributor/C+ and Paulo have been attempting to resolve the [bugs in the PR](https://github.com/Expensify/App/issues/16098#issuecomment-1579857826) but it's been a minute since the assignment. 
We've had a new C+ that might help with the review but my plan was to drop it into the external slack rooms if we didn't get movement this week.

**Comment by mallenexpensify** [comment]
Enjoy your week off @sophiepintoraetz .  checking on [here](https://expensify.slack.com/archives/C01SKUP7QR0/p1686612217047109)
Seems like a private #agency-issues channel could be a good idea

**Comment by Santhosh-Sellavel** [comment]
@tienifr
Can you give a summary of what happened on the issue, what's the reason this PR took this much time? Do we have anything unresolved here on the PR?

And I see there is some of the PR that has been inactive for almost the last 2 - 3 weeks. Daily updates are necessary unless even you are waiting for any input from CME/CPlus then bump them daily. Also if there is no response after multiple bumps. then please tag a BZ, so they can help the issue move forward, thanks!

**Comment by mallenexpensify** [comment]
I propose we have this issue worked on by an expert-contributor-agency. I propose we compensate @tienifr for the work completed so far (50% of the original $1000) price.   If we agree with this approach I can find someone to help next week.   @Santhosh-Sellavel @PauloGasparSv

**Comment by Santhosh-Sellavel** [comment]
Yeah, I agree!

**Comment by PauloGasparSv** [comment]
Also agree @mallenexpensify!

**Comment by robertKozik** [comment]
Hey :wave: I’m Robert Kozik from Software Mansion, an expert agency, and I would like to help close this issue out. I see that a PR by @tienifr is already posted, so here we are talking about closing out this one, not coming up with new?

**Comment by mallenexpensify** [comment]
@tienifr can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~015d2f9b2b3e902298

@robertKozik I assigned to you, thinking you'd propose a solution. Are you thinking something different?

**Comment by robertKozik** [comment]
Sure, let me familiarise with whole issue and I'll propose something. The proposal used in the PR looks promising though, if I would came up with the way of detecting modal open.

**Comment by robertKozik** [comment]
Hi! I've checked the issue to get some context and I think it will be ok to stick to the way proposed by @tienifr. Personally I'm not the fan of the document-scoped listeners, but in this case it's understandable. I'm thinking about changing the implementation though as keyDownAction can be under the `libs` directory. 

As I saw the main issue with this approach is how to determine whether or not composer is covered or not. I'm thinking about using the new navigation changes and base it on the route. As I checked, every open of RHN changes the url  i.e appends `/settings` to the report url. If we would check it, we will be sure that composer is not covered when path consist only of `/r/<reportID>` nothing else.

I need only to double check the assumptions about the RHN `path` case, if it's ok it should resolve the problem with the proposal and I'll document/implement it.

**Comment by robertKozik** [comment]
Hi! As for today I checked whether my idea with checking the url path is correct and it turned out better than I was thinking, with only 3 checkes it's possible to determine whether composer is covered or not (check the route, state of the popover menu and state of the emojiPickerPopover). I've created some rough implementation and checked the regressions which were listed inside the closed PR ([link](https://github.com/Expensify/App/pull/17649#issuecomment-1557577683)) and it looks like it already ticks all the boxes except the copy&paste one, which I wasn't aware of earlier.

All in all, I think that tomorrow I'll post PR for this.

**Comment by robertKozik** [comment]
I let myself write the proposal for this one, as the `copy&paste` function exceeds over the currently proposed implementation.
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
In order to access the composer it needs to be focused.
### What is the root cause of that problem?
After loosing focus on the composer it is not possible to access it again by keyboard.
### What changes do you think we should make in order to solve the problem?

**Handling key presses**
This will go in line with @tienifr proposal with few tweaks.
The state of the composer, whether it's covered or not, will be calculated in the new function: `checkComposerVisibility`. The function will calculate this state based on:
- current path with help of `ROUTES.parseReportRouteParams` (if reportID is not emptyString and isSubReportPage is false), 
- state of the EmojiPicker — here will be implemented new function `isEmojiPickerVisible` inside `EmojiPickerAction.js` file,
- state of the `+` popover — its accessible via `this.state.isMenuVisible`.

In the `libs/KeyboardShortcut` directory there will be new folder `KeyDownPressListener` (with different implementation for web and native) which will will manage document-scoped `keydown` listener. The listener will be added to the document when the composer is visible and removed when it's not (handle by `componentDidMount` and `componentWillUnmount` lifecycle methods).
Callback function for the listener will be `handleKeyDown` function inside `ReportActionCompose.js` file. This function will be responsible for handling key presses. It will be implemented in the following way:
- if the composer is not visible (based on `checkComposerVisibility`), it will do nothing,
- if the composer is visible, it will check if the pressed key is alphanumeric or not. If it's not, it will do nothing.
- if the pressed key is alphanumeric, it will check if we are not typing in the another input (f.e. search input). If we are, it will do nothing.
- if the pressed key is alphanumeric and we are not typing in the another input, it will call `this.replaceSelectionWithText` function (this is the function that is responsible for inserting text into the composer) with the pressed key as an argument.

**Handling special actions f.e. `copy&paste`**
Copy&paste cannot be handled inside `ReportActionCompose.js` file because it's not possible to parse the photos and text without duplicating the code from `Composer/index.js`. Because of that, the `paste` event will be handled inside `Composer/index.js` file. It will require to alter current implementation of listener for `paste` event. For now listener is hooked to TextInput component, but it will be moved to the document. In order to restrict this listener, same as keydown listener, the visibility of the composer will be passed to it from `ReportActionCompose` component. With this change, the `paste` event will be handled only when the composer is visible or focused (the check will be done inside `handlePaste` function in `Composer/index.js` file).
### What alternative solutions did you explore? (Optional)
1. For copy&paste I was considering to not change the textInput listener, but rather create the new document-scoped one along side. this document-scoped listener would be added or removed based on the current composer visibility. The problem with this solution is that it would require to add additional listener (we have many of them already). Moreover it would be harder to maintain, as the listener had to be removed on the propsUpdate.

**Comment by mallenexpensify** [comment]
@Santhosh-Sellavel , can you please review the proposal above?

**Comment by Santhosh-Sellavel** [comment]
@robertKozik So far looks good to me, let's see it in action!

**Comment by Santhosh-Sellavel** [comment]
Waiting for PR,  bump @robertKozik!

**Comment by robertKozik** [comment]
Actually I was just posting PR 😃

**Comment by PauloGasparSv** [comment]
Not overdue, https://github.com/Expensify/App/pull/21583 is up and under review

**Comment by Santhosh-Sellavel** [comment]
Review in progress

**Comment by robertKozik** [comment]
That's for sure regression from this PR. I'm on it to find fix for it

**Comment by sophiepintoraetz** [comment]
@Santhosh-Sellavel can you complete the Checklist please? I see you are raising requests through NewDot as well, so can you fire a payment through for $2000? thank you!

**Comment by Santhosh-Sellavel** [comment]
Please hold on from any action there @parasharrajat. I'm proposing to create follow-up issues to tackle this issue &invest some more time to test this to capture other bugs and handle them holistically by discussing and improving the feature handling.

cc: @robertKozik @sophiepintoraetz @PauloGasparSv

**Comment by sophiepintoraetz** [comment]
Thank you @Santhosh-Sellavel - that's a great idea. I will close this one out but let's reference this one in your master GH when you create it - lmk if you need a hand wrangling it together!

**Comment by kadiealexander** [comment]
Reopening to get some movement on these regressions:

https://github.com/Expensify/App/issues/23403
https://github.com/Expensify/App/issues/23184

@Santhosh-Sellavel please feel free to close once a new issue has been created as discussed [here](https://github.com/Expensify/App/issues/16098#issuecomment-1648754109).

**Comment by PauloGasparSv** [comment]
> Reopening to get some movement on these regressions:
> 
> #23403 #23184
> 
> @Santhosh-Sellavel please feel free to close once a new issue has been created as discussed [here](https://github.com/Expensify/App/issues/16098#issuecomment-1648754109).

We are discussing these in this [slack thread](https://expensify.slack.com/archives/C02NK2DQWUX/p1691069180339409?thread_ts=1690242585.218809&cid=C02NK2DQWUX)

**Comment by PauloGasparSv** [comment]
Not overdue, still in discussion over at the [thread](https://expensify.slack.com/archives/C02NK2DQWUX/p1691069180339409?thread_ts=1690242585.218809&cid=C02NK2DQWUX)

**Comment by PauloGasparSv** [comment]
Hey, we [decided over at slack](https://expensify.slack.com/archives/C02NK2DQWUX/p1691622467032029?thread_ts=1690242585.218809&cid=C02NK2DQWUX) to close this issue and work on the [2 bugs/regressions](https://github.com/Expensify/App/issues/16098#issuecomment-1667168533) in their respective issues instead of rolling the feature back and re-implementing it with a doc or fixing both regressions here.

I'll close this issue and comment this on each of the individual bug issues.

**Comment by JmillsExpensify** [comment]
Reviewed the details for @Santhosh-Sellavel. Approved for payment in NewDot based on the BZ summary above.

**Comment by sophiepintoraetz** [comment]
...is Melvin okay? I don't think that's correct, right @PauloGasparSv?

**Comment by Santhosh-Sellavel** [comment]
@sophiepintoraetz Yes Melvin is fine. This issue was a referenced in discussion or proposals
